### PR TITLE
WIP: Add password confirmation to Get-Credential

### PR DESCRIPTION
**Pull request by jboeshart**
# PR Summary

This PR adds a `-ConfirmPassword` parameter to `Get-Credential`. This helps to ensure that when entering a password, you really typed what you intended to (especially helpful with complex passwords). This only validates the text entered, and does not actually validate that the password meets any complexity requirements, nor does it actually validate it with any providers.

## PR Context

Originally logged as an enhancement request in #10625.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by jboeshart** [comment]
Still need to work on tests, and possibly documentation for the new parameter (need a little clarification on this).

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 1]
```diff
@@ -53,20 +57,24 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
             string message,
             string userName,
             string targetName,
+            bool confirmPassword,
             PSCredentialTypes allowedCredentialTypes,
             PSCredentialUIOptions options)
         {
             PSCredential cred = null;
             SecureString password = null;
+            SecureString confirmedPassword = null;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: It is not request to change: maybe `retypedPassword` would be more clear.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 2]
```diff
@@ -53,20 +57,24 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
             string userPrompt = null;
             string passwordPrompt = null;
+            string confirmPasswordPrompt = null;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: The same about name.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 3]
```diff
@@ -100,17 +108,80 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
+                confirmPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ConfirmPassword, userName);
+                bool passwordsMatch;
+                do
+                {
+                    WriteToConsole(passwordPrompt, true);
+                    password = ReadLineAsSecureString();
+                    WriteToConsole(confirmPasswordPrompt, true);
+                    confirmedPassword = ReadLineAsSecureString();
+                    
+                    IntPtr passwordBstr = IntPtr.Zero;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Please move the comparison code to private temporary method.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 4]
```diff
@@ -100,17 +108,80 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
+                        }
+                    }
+                    finally
+                    {
+                        if (confirmPasswordBstr != IntPtr.Zero)
+                            Marshal.ZeroFreeBSTR(confirmPasswordBstr);
+                        if (passwordBstr != IntPtr.Zero)
+                            Marshal.ZeroFreeBSTR(passwordBstr);
+                    }
+                } while (!passwordsMatch);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Will Ctrl-C works?

File: src/Microsoft.PowerShell.ConsoleHost/resources/ConsoleHostUserInterfaceSecurityResources.resx [hunk 5]
```diff
@@ -123,4 +123,10 @@
   <data name="PromptForCredential_Password" xml:space="preserve">
     <value>Password for user {0}: </value>
   </data>
+  <data name="PromptForCredential_ConfirmPassword" xml:space="preserve">
+    <value>Confirm password for user {0}: </value>
+  </data>
+    <data name="PasswordMismatch" xml:space="preserve">
+    <value>Passwords do not match. Enter your credentials. </value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/resources/ConsoleHostUserInterfaceSecurityResources.resx]: I am nor sure that I like the message but at least we should remove extra space.
```suggestion
    <value>Passwords do not match. Enter your credentials.</value>
```

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 6]
```diff
@@ -78,6 +78,19 @@ public string Title
         }
 
         private string _title = UtilsStrings.PromptForCredential_DefaultCaption;
+        
+        /// <summary>
+        /// Gets and sets whether to confirm the password.
+        /// </summary>
+        [Parameter(Mandatory = false)]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: It is default.
```suggestion
        [Parameter]
```

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 7]
```diff
@@ -78,6 +78,19 @@ public string Title
...
+        [ValidateNotNullOrEmpty]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: Why do we need this?

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 8]
```diff
@@ -78,6 +78,19 @@ public string Title
...
+        public SwitchParameter ConfirmPassword
+        {
+            get { return _confirmPassword; }
+            set { _confirmPassword = value; }
+        }
+
+        private bool _confirmPassword;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: Please use auto-implemented property.
```suggestion
        public SwitchParameter ConfirmPassword { get; set; }
```

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs [hunk 9]
```diff
@@ -306,7 +306,8 @@ public override
                         null,   // caption already written
                         null,   // message already written
                         null,
-                        string.Empty);
+                        string.Empty,
+                        false);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: Please use named arguments for all constants and nulls like
```suggestion
                        confirmPassword: false);
```

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs (hunk 4 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Yup, Ctrl-C works at any point in the process.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 6 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: Should I remove them from the other parameters as well? I only included it to align with the convention of the rest of the parameters, but I agree that it's not needed.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 7 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: Not needed, removed.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 6 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: No need fix other code without a request from maintainers. We prefer to separate style and functional changes by PRs/commits.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs (hunk 1 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: changed to `retpytedPassword`

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs [hunk 10]
```diff
@@ -306,7 +306,8 @@ public override
                         null,   // caption already written
                         null,   // message already written
                         null,
-                        string.Empty);
+                        string.Empty,
+                        confirmPassword: false);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: Please do the same for all arguments for consistency.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs (hunk 2 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: changed to `retypePasswordPrompt`

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs (hunk 3 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: I moved it to a private method in the class, not entirely sure what you're referring to around it being temporary. If the new private method isn't acceptable just let me know what we'd need to change.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs (hunk 10 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: Can do, but those were already there before I made any changes. Do you want me to include those changes here as well?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs (hunk 3 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Private method is that I meant. It simplify the code and allow us to replace comparison algorithm later if needed.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs (hunk 10 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: Yes, I see. My request is to get consistency. This makes a code more readable.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 11]
```diff
@@ -190,6 +143,59 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
+        {
+            IntPtr passwordBstr = IntPtr.Zero;
+            IntPtr retypedPasswordBstr = IntPtr.Zero;
+            int match = 0;
+            try
+            {
+                passwordBstr = Marshal.SecureStringToBSTR(password);
+                retypedPasswordBstr = Marshal.SecureStringToBSTR(retypedPassword);
+                int passwordLength = Marshal.ReadInt32(passwordBstr, -4);
+                int confirmPasswordLength = Marshal.ReadInt32(retypedPasswordBstr, -4);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: See answer in https://stackoverflow.com/questions/4502676/c-sharp-compare-two-securestrings-for-equality
It looks more reliable: 
```c#
        ss_bstr1_ptr = Marshal.SecureStringToBSTR(secureString1);
        ss_bstr2_ptr = Marshal.SecureStringToBSTR(secureString2);

        String str1 = Marshal.PtrToStringBSTR(ss_bstr1_ptr);
        String str2 = Marshal.PtrToStringBSTR(ss_bstr2_ptr);

        return str1.Equals(str2);
``

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 12]
```diff
@@ -100,17 +108,38 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
+            if (confirmPassword)
+            {
+                retypePasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ConfirmPassword, userName);
+                bool passwordsMatch;
+                do
+                {
+                    WriteToConsole(passwordPrompt, true);
+                    password = ReadLineAsSecureString();
+                    WriteToConsole(retypePasswordPrompt, true);
+                    retypedPassword = ReadLineAsSecureString();
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: If you look "else" block there is a null check that is Ctrl-C. I'd expect the same logic here too.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs (hunk 10 above)
Comment (jboeshart) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: So what do we want to do with this? Change it back to just `false` for consistency or leave it at `confirmPassword: false`?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 13]
```diff
@@ -100,17 +108,38 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
+
+                    passwordsMatch = ComparePasswords(password, retypedPassword);
+                    if (!passwordsMatch)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: In continue my previous comment I'd reverse the logic 
```c#
do 
{
    …
    if (passwd… == null)
    {
        return null;
    }

    …

    if (passwordsMatch)
    {
         return;
    }

    WriteToConsole(StringUtil.Format(ConsoleHostUserInterfaceSecurityResources…….
}
while (true);

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 14]
```diff
@@ -100,17 +108,38 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
+                    {
+                        WriteToConsole(StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PasswordMismatch), true);
+                        WriteLineToConsole();
+                    }
+                } while (!passwordsMatch);
+            }
+            else
             {
-                return null;
+                WriteToConsole(passwordPrompt, true);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Please use named argument for true.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs (hunk 10 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: Please add names for all, not only you changed, constants.

**Comment by iSazonov** [comment]
Replaced with #12782

**Pull request closed without merging**